File: ArraysAndHashing/Java/ValidAnagram/AnagramOptimizedSpace.java
/*
 * Time complexity: O(n)
 *
 * Space complexity:
 * - Output: O(1) - only a boolean is returned
 * - Auxiliary: O(1) - fixed-size int[26] array, since only lowercase English letters are used
 * - Total: O(1)
 *
 * The logic:
 * We use a fixed-size array of 26 integers to store the difference in letter counts between s and t.
 * We increment for each letter in s, decrement for each letter in t.
 * At the end, if all counts are zero, s and t are anagrams.
 * All extra space used is constant (does not grow with input size).
 */

package Java.ValidAnagram;

public class AnagramOptimizedSpace {
  public static boolean isAnagram(String s, String t) {
    if (s.length() != t.length())
      return false;

    int freq[] = new int[26];
    for (int i = 0; i < s.length(); i++) {
      freq[s.charAt(i) - 'a']++; // to map char to an index using ASCII and increment count in s
      freq[t.charAt(i) - 'a']--; // to map char to an index using ASCII and decrement count in t
    }

    return checkIfZero(freq);
  }

  // check if all frequencies are 0
  private static boolean checkIfZero(int[] freq) {
    for (int val : freq) {
      if (val != 0)
        return false;
    }
    return true;
  }

  public static void main(String[] args) {
    String s = "racecar";
    String t = "carrace";
    System.out.println(isAnagram(s, t));

    String s2 = "jar";
    String t2 = "jam";
    System.out.println(isAnagram(s2, t2));
  }

}

File: ArraysAndHashing/Python/ValidAnagram/anagram_optimized_space.py
'''
Time complexity: O(n)

Space complexity:
- Output: O(1) - only a boolean is returned
- Auxiliary: O(1) - fixed-size list of 26 integers (letter counts)
- Total: O(1)

The logic:
We use a fixed-size list (length 26) to track character counts.
We increment for each character in s and decrement for each in t.
If all frequencies are zero at the end, s and t are anagrams.
All extra space is constant, regardless of input size.
'''

class Solution:
  def isAnagram(self, s: str, t: str) -> bool:
    if len(s) != len(t): return False

    freq = [0] * 26;
    for i in range(0, len(s)):
      freq[ord(s[i]) - ord('a')] += 1
      freq[ord(t[i]) - ord('a')] -= 1

    return self.checkIfZero(freq)

  def checkIfZero(self, freq):
    for val in freq:
      if val != 0:
        return False
    return True

if __name__ == "__main__":
  s = "racecar"
  t = "carrace"

  s2 = "jar"
  t2 = "jam"

  obj = Solution()

  print(obj.isAnagram(s, t))
  print(obj.isAnagram(s2, t2))
